test/test_type_promotion.py

```diff
@@ -218,6 +218,47 @@ def test_create_bool_tensors(self):
         # this seems like odd behavior but ints also create float tensors, numpy doesn't have this function.
         self.assertEqual(torch.scalar_tensor(False, device=self.device), torch.tensor(0., device=self.device))
 
+    def test_comparison_ops_with_type_promotion(self):
+        value_for_type = {
+            torch.uint8: (1 << 5),
+            torch.int8: (1 << 5),
+            torch.int16: (1 << 10),
+            torch.int32: (1 << 20),
+            torch.int64: (1 << 30),
+            torch.float16: (1 << 10),
+            torch.float32: (1 << 20),
+            torch.float64: (1 << 30)
+        }
+        comparison_ops = [
+            dict(
+                name="lt",
+                out_op=lambda x, y, d: torch.lt(x, y, out=torch.empty(1, dtype=torch.bool, device=d)),
+                ret_op=lambda x, y: torch.lt(x, y),
+                compare_op=lambda x, y: x < y,
+            ),
+        ]
+        for device in torch.testing.get_all_device_types():
+            for op in comparison_ops:
+                for dt1 in torch.testing.get_all_math_dtypes(device):
+                    for dt2 in torch.testing.get_all_math_dtypes(device):
+                        val1 = value_for_type[dt1]
+                        val2 = value_for_type[dt2]
+                        t1 = torch.tensor([val1], dtype=dt1, device=device)
+                        t2 = torch.tensor([val2], dtype=dt2, device=device)
+                        expected = torch.tensor([op["compare_op"](val1, val2)], dtype=torch.bool)
+
+                        out_res = op["out_op"](t1, t2, device)
+                        self.assertEqual(out_res, expected)
+                        self.assertTrue(out_res.dtype == torch.bool)
+                        self.assertTrue(t1.dtype == dt1)
+                        self.assertTrue(t2.dtype == dt2)
+
+                        out_res = op["ret_op"](t1, t2)
+                        self.assertEqual(out_res, expected)
+                        self.assertTrue(out_res.dtype == torch.bool)
+                        self.assertTrue(t1.dtype == dt1)
+                        self.assertTrue(t2.dtype == dt2)
```

Comment:
The test case mentioned here: https://github.com/pytorch/pytorch/pull/25998#discussion_r327785865 appears to have disappeared, is that intentional?

Reply:
Previously it was the following behavior generated in LegacyTHFunctionsCPU.cpp:
```
case ScalarType::Byte: {
            auto result_ = checked_tensor_unwrap(result, "result", 0, "_th_lt_out", false, Backend::CPU, ScalarType::Bool);
            auto self_ = checked_tensor_unwrap(self, "self", 1, "_th_lt_out", false, Backend::CPU, ScalarType::Byte);
            auto other_ = other.toByte(); // Try to convert to tensor type
            THByteTensor_ltValue(result_, self_, other_);
            result_->maybe_zero_dim(self_->dim() == 0);
            return result;
            break;
        }
```
So when we try to convert scalar to the tensor type it will call toXXX() that will check https://github.com/pytorch/pytorch/blob/master/c10/core/Scalar.h#L60 conversion

Reply:
If I'm understanding you correctly, this PR has the following behavior:
```
import torch
x = torch.ones(3, dtype=torch.int)
x < 1.9
# tensor([False, False, False])
```
?

Reply:
One thing that would be really helpful, especially for people writing the release notes, is to populate the body of this pull request with exactly how we've changed semantics and what hasn't changed. This example would be a good example to comment about.

Reply:
**torch.lt(Tensor a, Tensor b)**
will compute common dtype (highest) based on inputs and then compare values. The result will be Bool tensor
```
>>> x = torch.tensor([0], dtype=torch.int)
>>> y = torch.tensor([0.5], dtype=torch.double)
>>> x < y
tensor([True])
```
**torch.lt(Tensor a, Tensor b, out=c)**
will compute common dtype (highest) based on inputs and then compare values. The result can be populated only to Bool tensor
```
>>> x = torch.tensor([0], dtype=torch.int)
>>> y = torch.tensor([0.5], dtype=torch.double)
>>> z = torch.empty([1], dtype=torch.bool)
>>> torch.lt(x, y, out=z)
tensor([True])
```
**a.lt_(Tensor b)**
Assumes that a and b has same dtype. Will not compute common dtype. Result will have type of a.
```
>>> x = torch.tensor([0], dtype=torch.double)
>>> y = torch.tensor([0.5], dtype=torch.double)
>>> x < y
tensor([True])
```
**torch.lt(Tensor a, Scalar b)**
will convert b to the same type as a, prior checking if there is no overflow.
```
>>> x = torch.tensor([0], dtype=torch.double)
>>> x < 0.5
tensor([True])

>>> x = torch.tensor([0], dtype=torch.int)
>>> x < 0.5
tensor([False])
```
**torch.lt(Tensor a, Scalar b, out=c)**
will convert b to the same type as a, prior checking if there is no overflow. The result can be populated only to Bool tensor
```
>>> x = torch.tensor([0], dtype=torch.double)
>>> torch.lt(x, 0.5, out=z)
tensor([True])
```
**torch.lt_(Tensor a, Scalar b)**
will convert b to the same type as a, prior checking if there is no overflow. Result will have type of a.
```
>>> x = torch.tensor([0], dtype=torch.int)
>>> x.lt_(1)
tensor([1], dtype=torch.int32)
>>> x = torch.tensor([0], dtype=torch.int)
>>> x.lt_(1.0)
tensor([1], dtype=torch.int32)
```

Reply:
You should toss that into your commit message and pull request body, not in a comment in the pull request review!

Reply:
Also it would be good to note which of these behaviors happened previously, and which of these behaviors is new, as of this PR.

Reply:
Moved to PR body

Reply:
Thank you, this is much clearer now